core/oracles/ethereum_oracle_verifier.go

```diff
@@ -0,0 +1,268 @@
+// duplicates but not result in a memory leak, agreed to postpone for now  (other verifiers have the same issue)
+
+func (s *EthereumOracleVerifier) ProcessEthereumContractCallResult(callEvent types.EthContractCallEvent) error {
+	// Possible that a single oracle trigger, where trigger is base off of vegatime, could result in chain events with
+	// different contents (and therefore different hashes) due to difference in the time when the validators query
+	// ethereum, it is assumed that the consumer of the oracle events will handle this.  Alternatively we could restrict
+	// triggers to use ethereum time and not allow vega time triggers, then no such issue.
+	if ok := s.ensureNotDuplicate(callEvent.Hash()); !ok {
+		s.log.Error("ethereum call event already exists",
+			logging.String("event", fmt.Sprintf("%+v", callEvent)))
+		return ErrDuplicatedEthereumCallEvent
+	}
+
+	pending := &pendingCallEvents{
+		callEvent: callEvent,
+		check:     func() error { return s.checkCallEventResult(callEvent) },
+	}
+
+	s.pendingCallEvents = append(s.pendingCallEvents, pending)
+
+	s.log.Info("ethereum call event received, starting validation",
+		logging.String("call-event", fmt.Sprintf("%+v", callEvent)))
+
+	// Timeout for the check set to 1 day, to allow for validator outage scenarios
+	return s.witness.StartCheck(
+		pending, s.onCallEventVerified, s.timeService.GetTimeNow().Add(24*time.Hour))
+}
+
+func (s *EthereumOracleVerifier) checkCallEventResult(contractCall types.EthContractCallEvent) error {
+	spec, err := s.ethCallSpecSource.GetSpec(contractCall.SpecId)
+	if err != nil {
+		return fmt.Errorf("failed to get call specification for id %s: %w", contractCall.SpecId, err)
+	}
+
+	blockHeight := &big.Int{}
+	blockHeight.SetUint64(contractCall.BlockHeight)
+	value, err := spec.Call(context.Background(), s.ethContractCaller, blockHeight)
+	if err != nil {
+		return fmt.Errorf("failed to execute call event spec: %w", err)
+	}
+
+	if !bytes.Equal(contractCall.Result, value) {
+		return fmt.Errorf("mismatched results for block %d", contractCall.BlockHeight)
+	}
+
+	if !spec.PassesFilters(contractCall.Result, contractCall.BlockHeight, contractCall.BlockTime) {
+		return fmt.Errorf("failed to pass filter check")
+	}
+
+	if err = s.ethConfirmations.Check(spec.RequiredConfirmations()); err != nil {
+		return fmt.Errorf("failed confirmations check: %w", err)
+	}
+
+	return nil
+}
+
+func (s *EthereumOracleVerifier) removePendingCallEvent(id string) error {
+	for i, v := range s.pendingCallEvents {
+		if v.GetID() == id {
+			s.pendingCallEvents = s.pendingCallEvents[:i+copy(s.pendingCallEvents[i:], s.pendingCallEvents[i+1:])]
+			return nil
+		}
+	}
+	return fmt.Errorf("invalid pending call event hash: %s", id)
+}
+
+func (s *EthereumOracleVerifier) onCallEventVerified(event interface{}, ok bool) {
+	pv, isPendingCallEvent := event.(*pendingCallEvents)
+	if !isPendingCallEvent {
+		s.log.Error("oracle verifier received invalid event")
```

Comment:
Shouldn't we return here instead of continuing?

Reply:
perhaps, I've based this code on what the stake verifier does, I'm guessing at this point there is little you can do, the signature of the callback doesn't have error IIRC

Reply:
If the type assertion fails `pv` will be nil and lower down `pv.callEvent` will panic.

Reply:
yeah, sorry I was mixing this up with the removal failure